pkg/server/local_observer_test.go

```diff
@@ -114,3 +109,71 @@ func TestLocalObserverServer_GetFlows(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, req.Number, uint64(i))
 }
+
+type testHook struct {
+	t            *testing.T
+	ciliumDaemon interface{}
+	stopEvery    int64
+
+	seenFlows int64
+}
+
+func (h *testHook) AfterServerInit(srv serveroption.Server) error {
+	assert.Equal(h.t, srv.GetOptions().CiliumDaemon, h.ciliumDaemon)
+	return nil
+}
+
+func (h *testHook) BeforeDecoding(ctx context.Context, payload *pb.Payload) (stop bool, err error) {
+	h.seenFlows++
+	assert.Equal(h.t, payload.Time.Seconds, h.seenFlows-1)
+	if h.seenFlows%h.stopEvery == 0 {
+		stop = true
+	}
+	return stop, nil
+}
+
+func (h *testHook) AfterDecoding(ctx context.Context, event *v1.Event) (stop bool, err error) {
+	assert.Equal(h.t, event.Timestamp.Seconds, h.seenFlows-1)
+	if h.seenFlows%h.stopEvery == 0 {
+		assert.Fail(h.t, "server did not stop decoding")
+	}
+	return stop, nil
+}
+
+func TestHooks(t *testing.T) {
+	numFlows := 10
+	queueSize := 0
+
+	type fakeCiliumDaemon struct{}
+	h := &testHook{
+		t:            t,
+		stopEvery:    2,
```

Comment:
where and how is this `% 2 == 0` covered in this test?

Reply:
https://github.com/cilium/hubble/blob/bbfe90eb6484a2a462b71e85fb6dc3d6c6311de9/pkg/server/local_observer_test.go#L137-L139

Reply:
I see.

I think adding some pure function hooks will make this a lot easier to comprehend as one test https://github.com/cilium/hubble/pull/193/files#r395140700

This also doesn't cover that two hooks A and B both doing `BeforeDecoding` get chained together correctly (first one returning stop).